torch/lib/THD/base/data_channels/DataChannelNccl.hpp

```diff
@@ -0,0 +1,232 @@
+  virtual ~DataChannelNccl();
+
+  bool init() override;
+  void destroy() override;
+
+  rank_type getRank() override;
+  rank_type getNumProcesses() override;
+
+  void allReduce(std::vector<at::Tensor>& input,
+                 std::vector<at::Tensor>& output,
+                 THDReduceOp operation,
+                 THDGroup = THDGroupWORLD) override;
+
+  void allReduce(at::Tensor& data,
+                 THDReduceOp operation,
+                 THDGroup groupId = THDGroupWORLD) override;
+
+  void allGather(std::vector<at::Tensor>& input,
+                 std::vector<at::Tensor>& output,
+                 THDGroup groupId = THDGroupWORLD) override;
+
+  void allGather(std::vector<at::Tensor>& output,
+                 at::Tensor& input,
+                 THDGroup groupId = THDGroupWORLD) override;
+
+  void reduce(std::vector<at::Tensor>& input,
+              THDReduceOp operation,
+              rank_type dstRank,
+              THDGroup groupId = THDGroupWORLD) override;
+
+  void reduce(at::Tensor& data,
+              THDReduceOp operation,
+              rank_type dstRank,
+              THDGroup groupId = THDGroupWORLD) override;
+
+  void broadcast(std::vector<at::Tensor>& data,
+                 rank_type srcRank,
+                 THDGroup groupId = THDGroupWORLD) override;
+
+  void broadcast(at::Tensor& data,
+                 rank_type srcRank,
+                 THDGroup groupId = THDGroupWORLD) override;
+
+  void barrier(THDGroup groupId = THDGroupWORLD) override;
+
+  THDGroup newGroup(const std::vector<rank_type>& ranks) override;
+
+  void destroyGroup(THDGroup groupId = THDGroupWORLD) override;
+
+  // Not supported functions
+  void gather(std::vector<at::Tensor>& output,
+              at::Tensor& input,
+              rank_type dstRank,
+              THDGroup groupId = THDGroupWORLD) override;
+
+  void scatter(std::vector<at::Tensor>& input,
+               at::Tensor& output,
+               rank_type srcRank,
+               THDGroup groupId = THDGroupWORLD) override;
+
+  void send(Scalar& data, rank_type dstRank) override;
+
+  void send(at::Tensor& data, rank_type dstRank) override;
+
+  void receive(Scalar& data, rank_type srcRank) override;
+
+  rank_type receive(at::Tensor& data) override;
+
+  void receive(at::Tensor& data, rank_type srcRank) override;
+
+  RequestNccl* isend(at::Tensor& data, rank_type dstRank) override;
+
+  RequestNccl* ireceive(at::Tensor& data, rank_type srcRank) override;
+
+private:
+
+  // Current process' rank
+  rank_type _rank;
+  // Number of processes in network
+  rank_type _numProcesses;
+
+  // Accept waiting timeout in milliseconds, optional
+  int _timeout;
+  // Master's address
+  std::string _masterAddr;
+  // Master's port
+  port_type _masterPort;
+  // Socket on which the master is listening
+  int _masterListeningSocket;
+  // Sockets on which the master is sending to each slave
+  std::vector<int> _masterSendingSockets;
```

Comment:
It would be useful to add a comment that these are not sorted by ranks and can be in arbitrary order